### PR TITLE
Add a new CoverageKind::Branch to MIR

### PR DESCRIPTION
**Pull request by Swatinem**
This adds a couple intermediate types, and threads branch coverage from MIR through to codegen, and emits branch coverage regions via LLVM.

Based on top of https://github.com/rust-lang/rust/pull/114399 and https://github.com/rust-lang/rust/pull/115058, and so far untested, as we do not (yet) generate any branches in MIR.

### CODE REVIEW

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 1]
```diff
@@ -491,6 +491,10 @@ impl<'tcx> Stable<'tcx> for mir::coverage::CoverageKind {
                     rhs: opaque(rhs),
                 }
             }
+            CoverageKind::Branch { true_op, false_op } => stable_mir::mir::CoverageKind::Branch {
+                true_op: opaque(true_op),
+                false_op: opaque(false_op),
+            },
```
Comment (Zalathar) [inline compiler/rustc_smir/src/rustc_smir/mod.rs:497]: Not directly related to this PR, but I was a little concerned to see that SMIR has copies of coverage-related data structures.

[I've opened a Zulip thread in #project-stable-mir](https://rust-lang.zulipchat.com/#narrow/stream/320896-project-stable-mir/topic/SMIR.20and.20StatementKind.3A.3ACoverage) to hopefully get some clarity on how SMIR and coverage should interact.

File: compiler/rustc_middle/src/mir/coverage.rs [hunk 2]
```diff
@@ -93,6 +93,10 @@ pub enum CoverageKind {
         op: Op,
         rhs: Operand,
     },
+    Branch {
+        true_op: Operand,
+        false_op: Operand,
+    },
```
Comment (Zalathar) [inline compiler/rustc_middle/src/mir/coverage.rs:99]: I think it's a little bit confusing to call these `true_op`/`false_op`, for two reasons:
- `Expression` uses the abbreviation `op` to mean *operation*, not operand
- Even though it makes sense to reuse the `Operand` type here, logically these aren't really “operands”

Perhaps `true_count`/`false_count`?

(In the coverage map parser for #114843 I just punted and called them `true_`/`false_`, which is another option.)

Comment (Zalathar) [reply compiler/rustc_middle/src/mir/coverage.rs:99]: Ah, I see you've already used the names `true_counter`/`false_counter` in the data structures that are prepared for FFI. That makes sense, but complicates things.

(Naming this stuff well is hard, especially when LLVM itself isn't very consistent about it!)

Comment (Swatinem) [reply compiler/rustc_middle/src/mir/coverage.rs]: I renamed half of these to `true_`/`false_`, and I agree that naming these is hard :-)

**Comment by Swatinem** [comment]
Rebased on top of https://github.com/rust-lang/rust/pull/114399 which makes the changes here a lot simpler.

**Comment by Swatinem** [comment]
A first proof of concept seems to be functional and it outputs the following counters / mappings (see https://github.com/rust-lang/rust/pull/115061/commits/7f985c43b7d3a75849761fb1badedd9472421fe6):

```
- Branch at (prev + 2, 8) to (start + 0, 12)
    true  = c1
    false = (c0 - c1)
```

As well as `llvm-cov` output:

```
  ------------------
  |  Branch (LL:13): [True: 1, False: 5]
  ------------------
```

There are still a couple of problems with this POC that I will take a look at this week:

- `code_region`s are not yet correct for macros and derives and match statements.
- `llvm-cov` is also outputting different lines for generic instantiation, but I believe that is rather the way how `llvm-cov` reports these things.
- The column numbers seem to be way off sometimes, reporting thing like `Branch (LL:192675): [True: 0, False: 0]`.
- This still needs to be hooked up to the existing expression simplification, as it is currently outputting `(Zero - Zero)` expressions for unreachable branches.
- It seems like the code is outputting duplicates for some `Code(Zero)` regions.

**Comment by Zalathar** [comment]
I was surprised to see changes to `coverage-dump`, because it should already work with branch regions as-is. (I tested it against a C program compiled with clang.)

In the existing implementation, it prints the raw true/false counters inline, and then also expands expressions (if any) on separate lines as needed.

File: src/tools/compiletest/src/runtest.rs [hunk 3]
```diff
@@ -723,7 +727,7 @@ impl<'test> TestCx<'test> {
 
     /// Replace line numbers in coverage reports with the placeholder `LL`,
     /// so that the tests are less sensitive to lines being added/removed.
-    fn anonymize_coverage_line_numbers(coverage: &str) -> Cow<'_, str> {
+    fn anonymize_coverage_line_numbers(line: &str) -> String {
```
Comment (Zalathar) [inline src/tools/compiletest/src/runtest.rs:730]: Note that this string is actually the whole coverage output, not just one line.

That's why we use `(?m:^)` to match the start of a line, and then use `Regex::replace_all` to modify all of the matching lines in one call.

File: compiler/rustc_session/src/config.rs [hunk 4]
```diff
@@ -168,6 +168,8 @@ pub enum MirSpanview {
 pub enum InstrumentCoverage {
     /// Default `-C instrument-coverage` or `-C instrument-coverage=statement`
     All,
+    /// Additionally, instrument branches and output branch overage
+    Branch,
```
Comment (Zalathar) [inline compiler/rustc_session/src/config.rs:172]: To save anyone else the trouble of checking whether this is accidentally stable (it's not):

New values of `InstrumentCoverage` are automatically treated as nightly-only unless they are explicitly added to the check shown below on line 2674, so `Branch` is correctly treated as a nightly-only value. 👍 

https://github.com/rust-lang/rust/blob/c6f5495591f1688b7d47a3c4085d84bd7dedde4c/compiler/rustc_session/src/config.rs#L2665-L2685

Comment (Swatinem) [reply compiler/rustc_session/src/config.rs:172]: Also, I haven’t yet hooked up the option, so it doesn’t do anything right now, yet.

**Comment by Swatinem** [comment]
> In the existing implementation, it prints the raw true/false counters inline, and then also expands expressions (if any) on separate lines as needed.

I didn’t know this was intentional, as the inline `Debug` output looks a bit untidy.
I reverted these changes, but renamed the identifiers to `r#true` to have slightly tidier output.

**Comment by davidtwco** [comment]
Use `@rustbot ready` when you want this to be reviewed.

**Comment by Dylan-DPC** [comment]
@Swatinem any updates on this? thanks

**Comment by Swatinem** [comment]
I will close this PR for now.
Smaller parts of it have landed independently, and @Zalathar is still hard at work overhauling the guts of coverage instrumentation, which I am keeping track of and reviewing. Apart from that I don’t have too much time to dedicate to this right now, though I might pick that up again once I do.

**Pull request closed without merging**